assets/src/edit-story/components/postLock/postLock.js

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
+import { useFeatures } from 'flagged';
+import { trackError } from '@web-stories-wp/tracking';
+
+/**
+ * Internal dependencies
+ */
+import { useAPI } from '../../app/api';
+import { useStory } from '../../app/story';
+import { useConfig } from '../../app/config';
+import { useCurrentUser } from '../../app/currentUser';
+import PostLockDialog from './postLockDialog';
+import PostTakeOverDialog from './postTakeOverDialog';
+
+function PostLock() {
+  const {
+    actions: { getStoryLockById, setStoryLockById, deleteStoryLockById },
+  } = useAPI();
+  const {
+    state: { currentUser },
+  } = useCurrentUser();
+  const {
+    storyId,
+    dashboardLink,
+    nonce: firstNonce,
+    postLock: { interval: postLockInterval, showLockedDialog },
+  } = useConfig();
+
+  const { previewLink, lockUser } = useStory(
+    ({
+      state: {
+        story: { previewLink, lockUser },
+      },
+    }) => ({
+      previewLink,
+      lockUser,
+    })
+  );
+
+  const { enablePostLocking } = useFeatures();
+  const [showDialog, setShowDialog] = useState(false);
+  const [isFirstTime, setIsFirstTime] = useState(true);
+  const [user, setUser] = useState({});
+  const [nonce, setNonce] = useState(firstNonce);
+
+  // When dialog is closed, then set current user to lock owner.
+  const closeDialog = useCallback(() => {
+    setShowDialog(false);
+    setStoryLockById(storyId);
+  }, [setStoryLockById, storyId]);
+
+  const currentUserLoaded = useMemo(
+    () => Boolean(Object.keys(currentUser).length),
+    [currentUser]
+  );
+
+  // When async call only if dialog is true, current user is loaded and post locking is enabled.
+  const doGetStoryLock = useCallback(() => {
```

Comment:
This function is really hard to grasp as it is not very readable and does a lot of things at once.

It might be beneficial to extract parts of it into an effect or something, or splitting this up into two functions.

Can be done in a follow-up PR of course, just wanted to mention it.

A downside of the current approach here is that we first have to do `GET` request to get the current lock and then a `POST` request, which seems a bit wasteful (compared to just 1 request to refresh an existing lock)

Reply:
I think this whole file is a little bit too big. I was thinking about breaking this into a hook / provider. We can do this later, as you say.